File: test_stream.py
import json
import sys
import urllib.request

SUPABASE_FUNCTION_URL = "https://zzlptwfqqnjhxtvmebqb.supabase.co/functions/v1/chat"
SUPABASE_ANON_KEY = "sb_publishable_8eu0QBwgFKoECWdlqf4DvQ_mtmVsixc"

def test_chat_streaming(prompt: str, model: str = "anacleto-32b", web_search: bool = False, deep_reasoning: bool = False):
    print("========================================================")
    print(" TESTING REAL-TIME STREAMING CHAT RESPONSE IN TERMINAL")
    print(f" Model: {model} | WebSearch: {web_search} | DeepReasoning: {deep_reasoning}")
    print(f" Prompt: '{prompt}'")
    print("========================================================\n")
    print("[STREAMING OUTPUT]: ", end="", flush=True)

    payload = json.dumps({
        "message": prompt,
        "attachment": "",
        "fileContent": "",
        "webSearch": web_search,
        "deepReasoning": deep_reasoning,
        "model": model,
        "history": []
    }).encode("utf-8")

    req = urllib.request.Request(
        SUPABASE_FUNCTION_URL,
        data=payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {SUPABASE_ANON_KEY}"
        },
        method="POST"
    )

    try:
        with urllib.request.urlopen(req) as response:
            buffer = ""
            for chunk in response:
                line = chunk.decode("utf-8")
                buffer += line
                
                if "\n\n" in buffer:
                    blocks = buffer.split("\n\n")
                    buffer = blocks.pop()
                    
                    for block in blocks:
                        event_type = ""
                        data_str = ""
                        for l in block.split("\n"):
                            if l.startswith("event: "):
                                event_type = l[7:].strip()
                            elif l.startswith("data: "):
                                data_str = l[6:].strip()
                        
                        if data_str:
                            try:
                                data = json.loads(data_str)
                                if event_type == "status":
                                    sys.stdout.write(f"\n[{data.get('message')}]\n")
                                    sys.stdout.flush()
                                elif event_type == "searchSummary":
                                    sys.stdout.write(f"\n[SEARCH SOURCES: {len(data.get('sources', []))}]\n")
                                    sys.stdout.flush()
                                elif event_type == "text" and "chunk" in data:
                                    sys.stdout.write(data["chunk"])
                                    sys.stdout.flush()
                                elif event_type == "done":
                                    sys.stdout.write(f"\n\n[STREAM COMPLETE - Engine: {data.get('model')} | Latency: {data.get('latency')}]\n")
                                    sys.stdout.flush()
                            except Exception:
                                pass

    except Exception as e:
        print(f"\n[STREAM ERROR]: {e}")

if __name__ == "__main__":
    prompt_input = sys.argv[1] if len(sys.argv) > 1 else "Explain recursion in programming in 3 simple bullet points."
    test_chat_streaming(prompt_input, model="anacleto-32b")

File: test_stream.ts
const NVIDIA_API_KEY = "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";

async function testStream() {
  console.log("Starting stream request...");
  const startTime = Date.now();

  try {
    const res = await fetch("https://integrate.api.nvidia.com/v1/chat/completions", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${NVIDIA_API_KEY}`
      },
      body: JSON.stringify({
        model: "nvidia/nemotron-3.5-lightning-30b-a3b",
        messages: [{ role: "user", content: "Explain quantum mechanics in one paragraph." }],
        temperature: 1,
        top_p: 0.95,
        max_tokens: 4096,
        stream: true,
        chat_template_kwargs: { enable_thinking: true },
        reasoning_budget: 4096
      })
    });

    if (!res.ok) throw new Error(`HTTP ${res.status}: ${await res.text()}`);

    let firstChunkTime = -1;
    let hasReasoning = false;
    
    // @ts-ignore
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      
      if (firstChunkTime === -1) {
        firstChunkTime = Date.now() - startTime;
        console.log(`\n[Time to first chunk: ${firstChunkTime}ms]`);
      }
      
      const chunkStr = decoder.decode(value, { stream: true });
      const lines = chunkStr.split('\n');
      
      for (const line of lines) {
        if (line.startsWith('data: ') && line !== 'data: [DONE]') {
          try {
            const data = JSON.parse(line.slice(6));
            const delta = data.choices[0]?.delta;
            
            if (delta?.reasoning_content) {
              hasReasoning = true;
              process.stdout.write(`\x1b[34m${delta.reasoning_content}\x1b[0m`);
            }
            if (delta?.content) {
              process.stdout.write(delta.content);
            }
          } catch(e) {}
        }
      }
    }
    
    console.log(`\n\nFinished in ${Date.now() - startTime}ms. Has reasoning: ${hasReasoning}`);
  } catch (err) {
    console.error("Stream failed:", err);
  }
}

testStream();
